### ISSUE TITLE
adjust AsnGenerator rules to avoid Level2b table creation for workflows that stop at Level1b

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#5655]
_Issue [JP-2495](https://jira.stsci.edu/browse/JP-2495) was created on JIRA by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam):_

As of DMS b7.7, certain instrument modes will only process through Level1b, and not proceed on to Level2a.  In some situations those datasets still get a Level2b association table written by AsnGenerator, which will never result in calling calibration because the requisite input Level2a files will never be generated.  The AsnGenerator rules should be adjusted to exclude creating Level2b asn tables for this particular instrument mode that will stop at Level1b:

EXP_TYPE = NRC_IMAGE  and 
DMS_NOTE = WFSC_LOS_JITTER

This should have been the requested behavior under JP-1872, but that ticket description was written incorrectly, so the rule change made under that ticket can be backed-out in favor of the one above.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554490#comment-554490):_

[Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) Do we have any pools with the new column included? If not, will any pool with NRC_TACQ exposures do to 'dummy up' a DMS_NOTE column to test changes?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554512#comment-554512):_

[Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) Yes, we have lots of examples of asn pools containing the DMS_NOTE column.  Here are some on Dstring2 from the nOPS11 test:
/ifs/archive/test/jwst/store/rspencer/tests/nOPS11/dtest2/JWSTDP-2021.4.0-27~de429a95/2021-09-21-190029/sdp/asn_creation/pools/

Every test suite would likely have some, so if you need some with certain values in them, we might be able to hunt some of those down by searching our database (we also put the asn pool content into a DB table).

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554582#comment-554582):_

Our regression tests covering association generation do not cover this connected pair of requirements, so I've moved a relevant pool file to an accessible directory for testing purposes.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=555136#comment-555136):_

Fixed by #6404

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=555185#comment-555185):_

Added jw00632_20210921t193255_pool.csv to Artifactory sdp/pools for regression testing.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554490#comment-554490):_

[Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) Do we have any pools with the new column included? If not, will any pool with NRC_TACQ exposures do to 'dummy up' a DMS_NOTE column to test changes?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554512#comment-554512):_

[Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) Yes, we have lots of examples of asn pools containing the DMS_NOTE column.  Here are some on Dstring2 from the nOPS11 test:
/ifs/archive/test/jwst/store/rspencer/tests/nOPS11/dtest2/JWSTDP-2021.4.0-27~de429a95/2021-09-21-190029/sdp/asn_creation/pools/

Every test suite would likely have some, so if you need some with certain values in them, we might be able to hunt some of those down by searching our database (we also put the asn pool content into a DB table).

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=554582#comment-554582):_

Our regression tests covering association generation do not cover this connected pair of requirements, so I've moved a relevant pool file to an accessible directory for testing purposes.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=555136#comment-555136):_

Fixed by #6404

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=555185#comment-555185):_

Added jw00632_20210921t193255_pool.csv to Artifactory sdp/pools for regression testing.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1872?focusedCommentId=566986#comment-566986):_

Given that JP-2495 has been created to fix the error implemented in this ticket, I don't think this should be reopened and included in the new work. The fix should be tracked in JP-2495 only.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567044#comment-567044):_

Fixed by #6681

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567507#comment-567507):_

I might need clarification on what associations are expected for the pool file you provided. When I run asn_generate on the pool, I get 14 associations, 7 Lv2ImageNonScience associations containing one member with exp_type NRC_TACQ, and 7 Lv2WFSC associations containing one member with exp_type NRC_IMAGE.

In the asn table link, I see only the seven Lv2WFSC associations - I presume those are not wanted, but are the Lv2ImageNonScience created? If not, should they be?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567522#comment-567522):_

Ah, sorry, I provided incomplete information.  The directory I listed only contained the association tables that were generated for the NRC_IMAGE, WFSC_LOS_JITTER cases of program 00761 that can't be built because the science data stops at Level-1b, and generates no Level-2a files to be used as inputs to these particular tables.

Other tables WERE generated for those programs and observations, and we did get proper association tables for the NRC_TACQ datasets that alllowed them to progress to Level-2b, as intended.  The only remaining deficiency is that we are getting tables for these NRC_IMAGE, WFSC_LOS_JITTER cases that we don't want, since we won't be able to generate Level-2b products for them.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567526#comment-567526):_

Understood.

I made a mistake on the hotfix this ticket asked for - the single-line change was only part of the fixes required. If the exp_type of the pool members to exclude associations for is NRC_IMAGE and not NRC_TACQ (as I knew), the new Constraint should have been moved from Lv2ImageNonScience to Lv2WFSC. I'll push a fix now.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567044#comment-567044):_

Fixed by #6681

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567467#comment-567467):_

As of the DMS b7.9 runs from 2022-01-23, we are still seeing cases where these settings are still resulting in Level2b (wfs_image2) association tables:

EXP_TYPE = NRC_IMAGE and
DMS_NOTE = WFSC_LOS_JITTER

e.g.,  machine tljwdmscsched2
Level-2b asn tables here:  ███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████

Pool file is here:
████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████

These pointings seem to have NRC_IMAGE and WFSC_LOS_JITTER in their pool file settings.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567507#comment-567507):_

I might need clarification on what associations are expected for the pool file you provided. When I run asn_generate on the pool, I get 14 associations, 7 Lv2ImageNonScience associations containing one member with exp_type NRC_TACQ, and 7 Lv2WFSC associations containing one member with exp_type NRC_IMAGE.

In the asn table link, I see only the seven Lv2WFSC associations - I presume those are not wanted, but are the Lv2ImageNonScience created? If not, should they be?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mike Swam](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mswam) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567522#comment-567522):_

Ah, sorry, I provided incomplete information.  The directory I listed only contained the association tables that were generated for the NRC_IMAGE, WFSC_LOS_JITTER cases of program 00761 that can't be built because the science data stops at Level-1b, and generates no Level-2a files to be used as inputs to these particular tables.

Other tables WERE generated for those programs and observations, and we did get proper association tables for the NRC_TACQ datasets that alllowed them to progress to Level-2b, as intended.  The only remaining deficiency is that we are getting tables for these NRC_IMAGE, WFSC_LOS_JITTER cases that we don't want, since we won't be able to generate Level-2b products for them.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2495?focusedCommentId=567526#comment-567526):_

Understood.

I made a mistake on the hotfix this ticket asked for - the single-line change was only part of the fixes required. If the exp_type of the pool members to exclude associations for is NRC_IMAGE and not NRC_TACQ (as I knew), the new Constraint should have been moved from Lv2ImageNonScience to Lv2WFSC. I'll push a fix now.
